src/ocamlorg_toplevel/bin/worker.ml

```diff
@@ -110,59 +121,50 @@ let execute =
     JsooTop.execute true ~pp_code ~highlight_location pp_result phrase;
     Format.pp_print_flush pp_code ();
     Format.pp_print_flush pp_result ();
-    let highlight = !highlighted in
-    let data =
-      Worker_rpc.create
-        ?highlight
-        ~stdout:(jstr_of_buffer stdout_buff)
-        ~stderr:(jstr_of_buffer stderr_buff)
-        ~sharp_ppf:(jstr_of_buffer code_buff)
-        ~caml_ppf:(jstr_of_buffer res_buff)
-        ()
-    in
-    highlighted := None;
-    let json = Worker_rpc.to_json data in
-    json
-
-let recv_from_page e =
-  let msg = (Message.Ev.data (Ev.as_type e) : Jv.t) in
-  match Jv.get msg "ty" |> Jv.to_string with
-  | "setup" ->
-    Js_of_ocaml.Sys_js.set_channel_flusher
-      stdout
-      (Buffer.add_string stdout_buff);
-    Js_of_ocaml.Sys_js.set_channel_flusher
-      stderr
-      (Buffer.add_string stderr_buff);
-    setup ();
-    setup_printers ();
-    let data =
-      Worker_rpc.create
-        ~stdout:(jstr_of_buffer stdout_buff)
-        ~stderr:(jstr_of_buffer stderr_buff)
-        ()
-    in
-    let json = Worker_rpc.to_json data in
-    Worker.G.post json
-  | "execute" ->
-    let phrase = Jv.get msg "phrase" |> Jv.to_string in
-    Worker.G.post (execute phrase)
-  | "complete" ->
-    let part = Jv.get msg "phrase" |> Jv.to_string in
-    let n, res = UTop_complete.complete ~phrase_terminator:";;" ~input:part in
-    let res =
-      List.filter
-        (fun (l, _) -> not (Astring.String.is_infix ~affix:"__" l))
-        res
-    in
-    let res = List.map fst res in
-    let o = Jv.obj [||] in
-    Jv.set o "n" (Jv.of_int n);
-    Jv.set o "l" (Jv.of_list Jv.of_string res);
-    let data = Worker_rpc.create ~completions:o () in
-    let json = Worker_rpc.to_json data in
-    Worker.G.post json
-  | _ ->
-    Worker.G.post (Worker_rpc.create ())
-
-let run () = Jv.(set global "onmessage" @@ Jv.repr recv_from_page)
+    IdlM.ErrM.return
+      Toplevel_api.
+        { stdout = buff_opt stdout_buff
+        ; stderr = buff_opt stderr_buff
+        ; sharp_ppf = buff_opt code_buff
+        ; caml_ppf = buff_opt res_buff
+        ; highlight = !highlighted
+        }
+
+let setup () =
+  Js_of_ocaml.Sys_js.set_channel_flusher stdout (Buffer.add_string stdout_buff);
+  Js_of_ocaml.Sys_js.set_channel_flusher stderr (Buffer.add_string stderr_buff);
+  setup ();
+  setup_printers ();
+  IdlM.ErrM.return
+    Toplevel_api.
+      { stdout = buff_opt stdout_buff
+      ; stderr = buff_opt stderr_buff
+      ; sharp_ppf = None
+      ; caml_ppf = None
+      ; highlight = None
+      }
+
+let complete phrase =
+  let n, res = UTop_complete.complete ~phrase_terminator:";;" ~input:phrase in
+  let res =
+    List.filter (fun (l, _) -> not (Astring.String.is_infix ~affix:"__" l)) res
```

Comment:
We could define `is_infix` with `String.sub` here to avoid adding the `Astring` library (even if it's already a transitive dep)